Collect systemd state and NeedReload state

tsipinakis opened the pull request.

tsipinakis committed:
Send systemd SystemState

tsipinakis committed:
Collect NeedDaemonReload property from all units

Sadly it seems systemd does not send PropertiesChanged notification for
unit properties so hard looping is the only way to do this.

tsipinakis committed:
Add config option to disable needreload scanning

tsipinakis committed:
Invert system state metric

tsipinakis committed:
Send only one value for NeedDaemonReload

luisfdez commented:
Hi @tsipinakis , @traylenator ,

I was giving it a try to the changes and I think there is an error case that could impact the execution of the plugin.

If after starting collectd, a "daemon-reexec" "daemon-reload" is executed, the systemd collectd plugin will stop working.

Could you take a look to that case?

What I see after the reexecs:
Error:
```
[2021-10-12 10:23:20] Unhandled python exception in read callback: DBusException: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
[2021-10-12 10:23:20] Traceback (most recent call last):
[2021-10-12 10:23:20]   File "/usr/lib/python2.7/site-packages/collectd_systemd.py", line 139, in read_callback
    self.send_system_state()
[2021-10-12 10:23:20]   File "/usr/lib/python2.7/site-packages/collectd_systemd.py", line 123, in send_system_state
    state = self.get_system_state()
[2021-10-12 10:23:20]   File "/usr/lib/python2.7/site-packages/collectd_systemd.py", line 68, in get_system_state
    state = self.manager_properties.Get('org.freedesktop.systemd1.Manager', 'SystemState', dbus_interface='org.freedesktop.DBus.Properties')
[2021-10-12 10:23:20]   File "/usr/lib64/python2.7/site-packages/dbus/proxies.py", line 145, in __call__
    **keywords)
[2021-10-12 10:23:20]   File "/usr/lib64/python2.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
[2021-10-12 10:23:20] DBusException: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
[2021-10-12 10:23:20] read-function of plugin `python.collectd_systemd' failed. Will suspend it for 120.000 seconds.
```
Subsequent executions fail as well:
```
[2021-10-12 10:25:20] Unhandled python exception in read callback: DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.0 was not provided by any .service files
[2021-10-12 10:25:20] Traceback (most recent call last):
[2021-10-12 10:25:20]   File "/usr/lib/python2.7/site-packages/collectd_systemd.py", line 139, in read_callback
    self.send_system_state()
[2021-10-12 10:25:20]   File "/usr/lib/python2.7/site-packages/collectd_systemd.py", line 123, in send_system_state
    state = self.get_system_state()
[2021-10-12 10:25:20]   File "/usr/lib/python2.7/site-packages/collectd_systemd.py", line 68, in get_system_state
    state = self.manager_properties.Get('org.freedesktop.systemd1.Manager', 'SystemState', dbus_interface='org.freedesktop.DBus.Properties')
[2021-10-12 10:25:20]   File "/usr/lib64/python2.7/site-packages/dbus/proxies.py", line 145, in __call__
    **keywords)
[2021-10-12 10:25:20]   File "/usr/lib64/python2.7/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
[2021-10-12 10:25:20] DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.0 was not provided by any .service files
[2021-10-12 10:25:20] read-function of plugin `python.collectd_systemd' failed. Will suspend it for 240.000 seconds.
```

tsipinakis committed:
Add try/catch around dbus calls

If systemd is re-execed the plugin loses connection with re-exec. Avoid
repeatedly failing by re-initializing the manager.

tsipinakis committed:
Retry once after failing to get system state/need reload

tsipinakis commented:
I haven't been able to reproduce this particular crash (in send_system_state) but I was able to get similar ones in the send_need_reload function.

I've added try/catch around all dbus calls, and made a init->retry flow similar to how units are handled in the main loop.
